greenlight/harness/testcase.py

```diff
@@ -19,10 +19,17 @@ def setUp(self, *args, **kwargs):
         MarionetteTestCase.setUp(self, *args, **kwargs)
         Puppeteer.set_marionette(self, self.marionette)
 
+        self._start_handle_count = len(self.marionette.window_handles)
         self.marionette.set_context('chrome')
         self.browser = self.windows.current
 
     def tearDown(self, *args, **kwargs):
-        MarionetteTestCase.tearDown(self, *args, **kwargs)
+        win_count = len(self.marionette.window_handles)
+        self.assertEqual(win_count, self._start_handle_count,
+                         "A test must not leak window handles. "
+                         "This test started the browser with %s open "
+                         "top level browsing contexts, but ended with %s." %
+                         (self._start_handle_count, win_count))
 
         self.prefs.restore_all_prefs()
+        MarionetteTestCase.tearDown(self, *args, **kwargs)
```

Comment:
There can always be cases a test accidentally doesn't close the window. It might not always be even the fault of the test, but just some underlying code has been thrown an assertion, and the line for closing all the remaining windows is not reached.

If there was really no assertion before and windows are still open, an extra assertion might be good but at the same time we should really close by default all the other open windows except the last browser window. With that action we can make the suite more foolproof by ensuring that the next test has the correct conditions and will not fail. In your current case all the windows will still kept open until a test closes them all or all tests are done, right?